lib/obj/src/context.rs

```diff
@@ -0,0 +1,64 @@
+use cranelift_codegen::isa::TargetFrontendConfig;
+use cranelift_entity::EntityRef;
+use cranelift_wasm::GlobalInit;
+use wasmtime_environ::{Module, VMOffsets};
+use wasmtime_runtime::{
+    SignatureRegistry, VMGlobalDefinition, VMSharedSignatureIndex, VMTableDefinition,
+};
+
+pub struct TableRelocation {
+    pub index: usize,
+    pub offset: usize,
+}
+
+pub fn layout_vmcontext(
+    module: &Module,
+    target_config: &TargetFrontendConfig,
+) -> (Box<[u8]>, Box<[TableRelocation]>) {
+    let ofs = VMOffsets::new(target_config.pointer_bytes(), &module);
+    let out_len = ofs.size_of_vmctx() as usize;
+    let mut out = Vec::with_capacity(out_len);
+    out.resize(out_len, 0);
+
+    let mut signature_registry = SignatureRegistry::new();
+    for (index, sig) in module.signatures.iter() {
+        let offset = ofs.vmctx_vmshared_signature_id(index) as usize;
+        let id = signature_registry.register(sig);
+        unsafe {
+            let to = out.as_mut_ptr().add(offset) as *mut VMSharedSignatureIndex;
+            (*to) = id;
```

Comment:
This should also use `ptr::write` and avoid depending on `VMSharedSignatureIndex`. If it would help, we can add a `TargetSharedSignatureIndex` type to `VMOffsets`, since it is fixed-size and the same on all platforms, and then add a test in lib/runtime/src/vmcontext.rs that asserts that it's the same size as the actual `VMSharedSignatureIndex`.